File: uni_modules/camera-view/utssdk/app-android/camera/ImageProcessor.kt
package uts.ste.camera

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import io.dcloud.uts.console
import java.io.File
import java.io.FileOutputStream

/**
 * 图片处理工具类
 */
object ImageProcessor {

    // 可配置的 JPEG 压缩质量
    var jpegQuality: Int = 90

    fun saveBitmapToFile(context: Activity?, bitmap: Bitmap, quality: Int = jpegQuality): String? {
        if (context == null) return null
        if (bitmap.isRecycled) {
            console.log("Bitmap 已被回收，无法保存")
            return null
        }
        return try {
            val cacheDir = context.cacheDir
            val fileName = "camera_${System.currentTimeMillis()}.jpg"
            val file = File(cacheDir, fileName)

            FileOutputStream(file).use { outputStream ->
                bitmap.compress(Bitmap.CompressFormat.JPEG, quality.coerceIn(1, 100), outputStream)
                outputStream.flush()
            }
            file.absolutePath
        } catch (e: Throwable) {
            console.log("保存图片失败：${e.message}")
            null
        }
    }

    fun rotateBitmap(bitmap: Bitmap, degrees: Float): Bitmap {
        if (degrees == 0f) return bitmap
        return try {
            val matrix = android.graphics.Matrix()
            matrix.postRotate(degrees)
            Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
        } catch (e: Throwable) {
            console.log("旋转图片失败：${e.message}")
            bitmap
        }
    }

    /**
     * 水平镜像翻转（用于前置摄像头自拍）
     */
    fun flipBitmapHorizontal(bitmap: Bitmap): Bitmap {
        return try {
            val matrix = android.graphics.Matrix()
            matrix.preScale(-1f, 1f)
            Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
        } catch (e: Throwable) {
            console.log("镜像翻转失败：${e.message}")
            bitmap
        }
    }

    /**
     * 将水印配置叠加绘制到照片上
     *
     * @param context          Activity（获取 displayMetrics）
     * @param bitmap           原始照片 Bitmap
     * @param watermarks       水印配置列表
     * @param containerWidthDp 相机容器宽度（dp）；为 0 时退回到屏幕宽度（向后兼容）
     */
    fun drawWatermarks(context: Activity, bitmap: Bitmap, watermarks: List<Map<String, Any?>>,
                       containerWidthDp: Int = 0): Bitmap {
        if (watermarks.isEmpty()) return bitmap

        val mutableBitmap = bitmap.copy(Bitmap.Config.ARGB_8888, true)
        val canvas = Canvas(mutableBitmap)

        val density = context.resources.displayMetrics.density
        val photoWidth = mutableBitmap.width.toFloat()

        // dp → photo-px 比例：
        //   正确公式：photoWidth / containerWidthDp（水印 dp 坐标直接对应容器 dp 宽度）
        //   旧公式（全屏）：density * (photoWidth / screenWidthPx) ≈ photoWidth / (screenWidthPx/density)
        //   两者在容器铺满屏幕时等价；容器非全屏时旧公式偏移，改为正确公式。
        val containerDp = if (containerWidthDp > 0) containerWidthDp.toFloat()
                          else (context.resources.displayMetrics.widthPixels / density)
        val dpToPhoto = photoWidth / containerDp

        for (config in watermarks) {
            val type = config["type"] as? String ?: continue
            val style = config["style"] as? Map<*, *>

            when (type) {
                "text" -> drawTextWatermark(canvas, config, style, dpToPhoto, photoWidth)
                "image" -> drawImageWatermark(canvas, config, style, dpToPhoto)
            }
        }
        return mutableBitmap
    }

    private fun drawTextWatermark(
        canvas: Canvas,
        config: Map<String, Any?>,
        style: Map<*, *>?,
        dpToPhoto: Float,
        photoWidth: Float
    ) {
        val text = config["text"] as? String ?: return
        val paint = Paint(Paint.ANTI_ALIAS_FLAG)

        // 字体样式
        val fontSizeDp = (style?.get("fontSize") as? Number)?.toFloat() ?: 14f
        paint.textSize = fontSizeDp * dpToPhoto

        val fontColor = style?.get("fontColor") as? String ?: "#FFFFFF"
        paint.color = try { Color.parseColor(fontColor) } catch (e: Exception) { Color.WHITE }

        val fontWeight = style?.get("fontWeight") as? String ?: "normal"
        val isBold = fontWeight == "bold" || (fontWeight.toIntOrNull() ?: 0) >= 700
        val fontStyleVal = style?.get("fontStyle") as? String ?: "normal"
        val isItalic = fontStyleVal == "italic"
        paint.typeface = when {
            isBold && isItalic -> Typeface.create(Typeface.DEFAULT, Typeface.BOLD_ITALIC)
            isBold -> Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
            isItalic -> Typeface.create(Typeface.DEFAULT, Typeface.ITALIC)
            else -> Typeface.DEFAULT
        }

        // 位置
        val topDp = (style?.get("top") as? Number)?.toFloat() ?: 0f
        val topPx = topDp * dpToPhoto
        val fm = paint.fontMetrics
        val y = topPx - fm.ascent

        val textAlign = style?.get("textAlign") as? String ?: "left"
        val textOffsetDp = (style?.get("textOffset") as? Number)?.toFloat() ?: 0f
        val textOffsetPx = textOffsetDp * dpToPhoto

        val x = when (textAlign) {
            "center" -> { paint.textAlign = Paint.Align.CENTER; photoWidth / 2f + textOffsetPx }
            "right" -> { paint.textAlign = Paint.Align.RIGHT; photoWidth - textOffsetPx }
            else -> { paint.textAlign = Paint.Align.LEFT; textOffsetPx }
        }

        canvas.drawText(text, x, y, paint)
    }

    private fun drawImageWatermark(
        canvas: Canvas,
        config: Map<String, Any?>,
        style: Map<*, *>?,
        dpToPhoto: Float
    ) {
        val imageUrl = config["image"] as? String ?: return

        val watermarkBitmap: Bitmap? = CameraViewManager.getPreloadedBitmap(imageUrl)
            ?: run {
                if (!imageUrl.startsWith("http")) {
                    try {
                        // decodeFile 只接受纯路径，去掉可能携带的 file:// 前缀
                        val path = if (imageUrl.startsWith("file://")) imageUrl.removePrefix("file://") else imageUrl
                        BitmapFactory.decodeFile(path)
                    } catch (e: Exception) {
                        console.log("本地水印图片加载失败：${e.message}"); null
                    }
                } else {
                    console.log("水印图片未预加载，跳过绘制：$imageUrl"); null
                }
            }
        watermarkBitmap ?: return

        val rawWidthDp = (style?.get("width") as? Number)?.toFloat()
        val rawHeightDp = (style?.get("height") as? Number)?.toFloat()
        val topPx = ((style?.get("top") as? Number)?.toFloat() ?: 0f) * dpToPhoto
        val leftPx = ((style?.get("left") as? Number)?.toFloat() ?: 0f) * dpToPhoto

        val bmpW = watermarkBitmap.width.toFloat()
        val bmpH = watermarkBitmap.height.toFloat()

        // 根据传入的宽高决定目标尺寸：
        //   - 都未传：使用图片实际像素尺寸（不做缩放）
        //   - 都传了：拉伸到指定尺寸
        //   - 只传宽：高按比例缩放
        //   - 只传高：宽按比例缩放
        val dstWidthPx: Float
        val dstHeightPx: Float
        when {
            rawWidthDp != null && rawHeightDp != null -> {
                // 宽高都传 → 拉伸
                dstWidthPx = rawWidthDp * dpToPhoto
                dstHeightPx = rawHeightDp * dpToPhoto
            }
            rawWidthDp != null -> {
                // 只传宽 → 高按比例
                dstWidthPx = rawWidthDp * dpToPhoto
                dstHeightPx = if (bmpW > 0f) dstWidthPx * bmpH / bmpW else dstWidthPx
            }
            rawHeightDp != null -> {
                // 只传高 → 宽按比例
                dstHeightPx = rawHeightDp * dpToPhoto
                dstWidthPx = if (bmpH > 0f) dstHeightPx * bmpW / bmpH else dstHeightPx
            }
            else -> {
                // 都未传 → 使用图片实际像素尺寸
                dstWidthPx = bmpW
                dstHeightPx = bmpH
            }
        }

        val scaledBitmap = Bitmap.createScaledBitmap(
            watermarkBitmap,
            dstWidthPx.toInt().coerceAtLeast(1),
            dstHeightPx.toInt().coerceAtLeast(1),
            true
        )
        canvas.drawBitmap(scaledBitmap, leftPx, topPx, null)
        // 回收缩放后的 Bitmap（createScaledBitmap 创建了新对象）
        if (!scaledBitmap.isRecycled) {
            scaledBitmap.recycle()
        }
    }
}

File: uni_modules/camera-view/utssdk/app-ios/camera/ImageProcessor.swift
import UIKit
import Foundation

/// 图片处理工具类（对标 Android ImageProcessor.kt）
/// 单例模式，负责：水印绘制、图片旋转/镜像、保存到缓存目录
@objc public class ImageProcessor: NSObject {

    // MARK: - 单例
    @objc public static let shared = ImageProcessor()
    private override init() { super.init() }

    // MARK: - JPEG 压缩质量（可配置）
    @objc public var jpegQuality: CGFloat = 0.9

    // MARK: - 保存图片到缓存目录

    /// 将 UIImage 保存为 JPEG 到 App 缓存目录，返回文件绝对路径
    @objc public func saveImageToCache(_ image: UIImage, quality: CGFloat = -1) -> String? {
        let q = quality < 0 ? jpegQuality : min(max(quality, 0.01), 1.0)
        NSLog("[ImageProcessor] saveImageToCache: image.size=\(image.size), imageOrientation=\(image.imageOrientation)")
        guard let data = image.jpegData(compressionQuality: q) else {
            NSLog("[ImageProcessor] saveImageToCache: jpegData compression failed")
            return nil
        }
        NSLog("[ImageProcessor] saveImageToCache: jpeg data length=\(data.count) bytes, quality=\(q)")
        let cachesDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first
        guard let dir = cachesDir else { return nil }
        let fileName = "camera_\(Int(Date().timeIntervalSince1970 * 1000)).jpg"
        let fileURL = dir.appendingPathComponent(fileName)
        do {
            try data.write(to: fileURL)
            NSLog("[ImageProcessor] saveImageToCache: saved to \(fileURL.path)")
            return fileURL.path
        } catch {
            NSLog("[ImageProcessor] saveImageToCache: write failed \(error.localizedDescription)")
            return nil
        }
    }

    // MARK: - 图片旋转

    /// 按指定角度旋转图片（角度：0 / 90 / 180 / 270）
    @objc public func rotateImage(_ image: UIImage, degrees: CGFloat) -> UIImage {
        if degrees == 0 { return image }
        let radians = degrees * .pi / 180
        let newSize: CGSize
        // 90° / 270° 时宽高互换
        if Int(degrees) % 180 != 0 {
            newSize = CGSize(width: image.size.height, height: image.size.width)
        } else {
            newSize = image.size
        }
        UIGraphicsBeginImageContextWithOptions(newSize, false, image.scale)
        guard let context = UIGraphicsGetCurrentContext() else { return image }
        context.translateBy(x: newSize.width / 2, y: newSize.height / 2)
        context.rotate(by: radians)
        image.draw(in: CGRect(
            x: -image.size.width / 2,
            y: -image.size.height / 2,
            width: image.size.width,
            height: image.size.height
        ))
        let rotated = UIGraphicsGetImageFromCurrentImageContext() ?? image
        UIGraphicsEndImageContext()
        return rotated
    }

    // MARK: - 水平镜像翻转（用于前置摄像头）

    @objc public func flipImageHorizontal(_ image: UIImage) -> UIImage {
        UIGraphicsBeginImageContextWithOptions(image.size, false, image.scale)
        guard let context = UIGraphicsGetCurrentContext() else { return image }
        context.translateBy(x: image.size.width, y: 0)
        context.scaleBy(x: -1, y: 1)
        image.draw(in: CGRect(origin: .zero, size: image.size))
        let flipped = UIGraphicsGetImageFromCurrentImageContext() ?? image
        UIGraphicsEndImageContext()
        return flipped
    }

    // MARK: - 绘制水印

    /// 将水印配置列表叠加绘制到照片上，返回新图片
    /// - Parameters:
    ///   - image: 原始照片
    ///   - watermarks: 水印配置数组（与 Android 端格式完全一致）
    @objc public func drawWatermarks(on image: UIImage, watermarks: [[String: Any]]) -> UIImage {
        guard !watermarks.isEmpty else { return image }

        UIGraphicsBeginImageContextWithOptions(image.size, false, image.scale)
        image.draw(in: CGRect(origin: .zero, size: image.size))

        let photoW = image.size.width
        let photoH = image.size.height

        // 以屏幕主 scale 与图片宽度计算 dp→px 的比例
        // 与 Android 端 dpToPhoto = density * (photoWidth / screenWidthPx) 等效
        let screenScale = UIScreen.main.scale
        let screenW = UIScreen.main.bounds.width * screenScale
        let dpToPhoto = screenScale * (photoW / screenW)

        for config in watermarks {
            guard let type = config["type"] as? String else { continue }
            let style = config["style"] as? [String: Any]

            switch type {
            case "text":
                drawTextWatermark(config: config, style: style, dpToPhoto: dpToPhoto, photoW: photoW, photoH: photoH)
            case "image":
                drawImageWatermark(config: config, style: style, dpToPhoto: dpToPhoto)
            default:
                break
            }
        }

        let result = UIGraphicsGetImageFromCurrentImageContext() ?? image
        UIGraphicsEndImageContext()
        return result
    }

    // MARK: - 绘制文字水印

    private func drawTextWatermark(
        config: [String: Any],
        style: [String: Any]?,
        dpToPhoto: CGFloat,
        photoW: CGFloat,
        photoH: CGFloat
    ) {
        guard let text = config["text"] as? String else { return }

        // 字体大小
        let fontSizeDp = (style?["fontSize"] as? CGFloat) ?? 14
        let fontSize = fontSizeDp * dpToPhoto

        // 字体粗细 / 斜体
        let fontWeight = style?["fontWeight"] as? String ?? "normal"
        let fontStyleVal = style?["fontStyle"] as? String ?? "normal"
        let isBold   = fontWeight == "bold" || (Int(fontWeight) ?? 0) >= 700
        let isItalic = fontStyleVal == "italic"

        var traits: UIFontDescriptor.SymbolicTraits = []
        if isBold   { traits.insert(.traitBold) }
        if isItalic { traits.insert(.traitItalic) }

        let font: UIFont
        if traits.isEmpty {
            font = UIFont.systemFont(ofSize: fontSize)
        } else if let descriptor = UIFont.systemFont(ofSize: fontSize).fontDescriptor.withSymbolicTraits(traits) {
            font = UIFont(descriptor: descriptor, size: fontSize)
        } else {
            font = UIFont.systemFont(ofSize: fontSize)
        }

        // 字体颜色
        let fontColor = parseColor(style?["fontColor"] as? String) ?? .white

        // 位置
        let topDp  = (style?["top"] as? CGFloat) ?? 0
        let topPx  = topDp * dpToPhoto

        let textAlignStr = style?["textAlign"] as? String ?? "left"
        let textOffsetDp = (style?["textOffset"] as? CGFloat) ?? 0
        let textOffsetPx = textOffsetDp * dpToPhoto

        // 计算文字尺寸
        let attrs: [NSAttributedString.Key: Any] = [
            .font: font,
            .foregroundColor: fontColor
        ]
        let textSize = (text as NSString).size(withAttributes: attrs)

        let x: CGFloat
        let paragraphStyle = NSMutableParagraphStyle()
        switch textAlignStr {
        case "center":
            x = photoW / 2 - textSize.width / 2 + textOffsetPx
            paragraphStyle.alignment = .center
        case "right":
            x = photoW - textSize.width - textOffsetPx
            paragraphStyle.alignment = .right
        default:
            x = textOffsetPx
            paragraphStyle.alignment = .left
        }

        let allAttrs: [NSAttributedString.Key: Any] = [
            .font: font,
            .foregroundColor: fontColor,
            .paragraphStyle: paragraphStyle
        ]

        let drawRect = CGRect(x: x, y: topPx, width: textSize.width, height: textSize.height)
        (text as NSString).draw(in: drawRect, withAttributes: allAttrs)
    }

    // MARK: - 绘制图片水印

    private func drawImageWatermark(
        config: [String: Any],
        style: [String: Any]?,
        dpToPhoto: CGFloat
    ) {
        guard let imageUrl = config["image"] as? String else { return }

        // 优先取预加载的图片，否则尝试本地加载（与 Android 端一致）
        var watermarkImage: UIImage? = CameraViewManager.shared.getPreloadedImage(for: imageUrl)
        if watermarkImage == nil {
            if !imageUrl.hasPrefix("http") {
                watermarkImage = UIImage(contentsOfFile: imageUrl)
                if watermarkImage == nil {
                    return
                }
            } else {
                // 网络图片未预加载，跳过（与 Android 端一致）
                return
            }
        }
        guard let wImg = watermarkImage else { return }

        let rawWidthDp  = style?["width"]  as? CGFloat
        let rawHeightDp = style?["height"] as? CGFloat
        let topPx  = ((style?["top"]  as? CGFloat) ?? 0) * dpToPhoto
        let leftPx = ((style?["left"] as? CGFloat) ?? 0) * dpToPhoto

        let bmpW = wImg.size.width
        let bmpH = wImg.size.height

        // 目标尺寸（逻辑与 Android 端完全一致）
        let dstW: CGFloat
        let dstH: CGFloat
        if let rw = rawWidthDp, let rh = rawHeightDp {
            dstW = rw * dpToPhoto
            dstH = rh * dpToPhoto
        } else if let rw = rawWidthDp {
            dstW = rw * dpToPhoto
            dstH = bmpW > 0 ? dstW * bmpH / bmpW : dstW
        } else if let rh = rawHeightDp {
            dstH = rh * dpToPhoto
            dstW = bmpH > 0 ? dstH * bmpW / bmpH : dstH
        } else {
            dstW = bmpW
            dstH = bmpH
        }

        let drawRect = CGRect(x: leftPx, y: topPx, width: max(dstW, 1), height: max(dstH, 1))
        wImg.draw(in: drawRect)
    }

    // MARK: - 颜色解析（与 CameraViewManager 中的 parseColor 保持一致）

    private func parseColor(_ hexStr: String?) -> UIColor? {
        guard var hex = hexStr else { return nil }
        hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        if hex.count == 6 { hex = "FF" + hex }
        guard hex.count == 8 else { return nil }
        var value: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&value)
        let a = CGFloat((value >> 24) & 0xFF) / 255
        let r = CGFloat((value >> 16) & 0xFF) / 255
        let g = CGFloat((value >>  8) & 0xFF) / 255
        let b = CGFloat( value        & 0xFF) / 255
        return UIColor(red: r, green: g, blue: b, alpha: a)
    }
}
